support fetching original package names and import paths via reflection

mvdan opened the pull request:
(see commit messages - please do not squash)

Updates #849.

lu4p commented on reflect.go:
```diff
@@ -45,6 +46,14 @@ func (ri *reflectInspector) recordReflection(ssaPkg *ssa.Package) {
 	// if a new reflectAPI is found we need to Re-evaluate all functions which might be using that API
 	newDone := len(ri.result.ReflectAPIs) + len(ri.result.ReflectObjectNames)
 	if newDone > prevDone {
+		// At least one type in this package is used with reflection;
+		// the package name and import path are reachable via reflection as well.
+		// Note that package main is compiled as having import path "main" as well.
+		ri.result.ReflectObjectNames[ri.lpkg.obfuscatedPackageName()] = ri.lpkg.Name
+		if ri.lpkg.Name != "main" {
+			ri.result.ReflectObjectNames[ri.lpkg.obfuscatedImportPath()] = ri.lpkg.ImportPath
+		}
+
```
I think this is too broad, reflection is very common I expect this would exclude most import paths of real world packages from obfuscation. 

Maybe instead hide this behavior behind a flag like `-includePaths`, or make it opt-in on a package level via a magic comment/ constant. Or maybe even add the ability to exclude specific packages from obfuscation.

mvdan replied:
Yes, this would lower the quality of obfuscation in some cases, but I strongly believe that the default behavior should be for any Go package to work out of the box. Stronger obfuscation which may break some code should be opt-in, not opt-out. That is better UX - stronger obfuscation is a bit of a power user feature - and should also prevent us from getting dozens more "garble breaks with package X" issues filed when the problem tends to be related to reflection.

So, yes, we can and should add some way to opt into stronger obfuscation even when reflection is used. I would suggest that we tackle that in a follow-up PR because it would be a slightly involved change. I lean towards some sort of directive comment like you suggest.

lu4p replied:
I think this is very similar to giving up on import path obfuscation entirely. 

Other than wails I think there has been no one who's programs break due to import path obfuscation, so I think this is a rare edge case and should be opt-in.

mvdan replied:
Removed for now; see https://github.com/burrowers/garble/pull/904#issuecomment-2585711395.

lu4p commented on reflect.go:
```diff
@@ -464,44 +473,43 @@ func recordedObjectString(obj types.Object) objectString {
+// obfuscatedObjectName returns the obfucated name of a types.Object,
 // parent is needed to correctly get the obfucated name of struct fields
-func reflectedObjectString(obj types.Object, parent *types.Struct) string {
-	if obj == nil {
-		return ""
-	}
```
I think that the obfuscated name of a nil obj is `""` is fine. I don't see a benefit of the callers having to make sure obj is non-nil.

mvdan replied:
I don't think this is a big deal either way, but I personally disagree - a nil obj is an edge case that we should break out of as soon as reasonable, rather than keep going down the logic. All other callers already ensure that obj is non-nil, it was just two that were careless. If any code path ends up with a nil obj, that's possibly a bug that we need to investigate and fix - rather than silently assuming that the result is an empty string.

lu4p commented on reflect.go:
```diff
@@ -189,13 +198,13 @@ func (ri *reflectInspector) checkFunction(fun *ssa.Function) {
 			switch inst := inst.(type) {
 			case *ssa.Store:
 				obj := typeToObj(inst.Addr.Type())
-				if usedForReflect(ri.result, obj) {
+				if obj != nil && ri.usedForReflect(ri.result, obj) {
```
revert

lu4p commented on reflect.go:
```diff
@@ -189,13 +198,13 @@ func (ri *reflectInspector) checkFunction(fun *ssa.Function) {
 			switch inst := inst.(type) {
 			case *ssa.Store:
 				obj := typeToObj(inst.Addr.Type())
-				if usedForReflect(ri.result, obj) {
+				if obj != nil && ri.usedForReflect(ri.result, obj) {
 					ri.recordArgReflected(inst.Val, make(map[ssa.Value]bool))
 					ri.propagatedInstr[inst] = true
 				}
 			case *ssa.ChangeType:
 				obj := typeToObj(inst.X.Type())
-				if usedForReflect(ri.result, obj) {
+				if obj != nil && ri.usedForReflect(ri.result, obj) {
```
revert

lu4p commented on testdata/script/reflect.txtar:
```diff
@@ -2,7 +2,7 @@ exec garble build
 cmp stdout main.stdout
 
-! binsubstr main$exe 'garble_main.go' 'test/main' 'importedpkg.' 'DownstreamObfuscated' 'SiblingObfuscated' 'IndirectObfuscated' 'IndirectNamedWithoutReflect' 'AliasIndirectNamedWithReflect' 'AliasIndirectNamedWithoutReflect' 'FmtTypeField' 'LocalObfuscated'
+! binsubstr main$exe 'garble_main.go' 'DownstreamObfuscated' 'SiblingObfuscated' 'IndirectObfuscated' 'IndirectNamedWithoutReflect' 'AliasIndirectNamedWithReflect' 'AliasIndirectNamedWithoutReflect' 'FmtTypeField' 'LocalObfuscated'
```
I would like to see `importedpkg` replaced by `indirect` to check that only certain import paths are excluded from obfuscation. Also check that `importedpkg` can be found in the binary with `binsubstr`.

mvdan replied:
yeah it's a fair point - I should update this PR to ensure that package import paths unrelated to reflection are excluded from the binary.

lu4p replied:
I meant that once this is opt-in you show that one packages import path is obfuscated while another's is not although both packages use reflection.

lu4p reviewed (changes requested):
I think this PR is too aggressive at including import paths.

mvdan commented:
Given that we don't agree on this, and that the approach as written doesn't fix Wails anyway, I'll need to give this a second attempt in a separate PR.

For now, I have updated the PR to remove package name and path deobfuscation, and fixed a failing test as well. All of the changes leading up to the original reflect change are still valuable in my opinion, and should be merged.
